### PR TITLE
Add necessary to extract v3 FIPs

### PR DESCRIPTION
**Pull request by superna9999**
The V3 FIPs are generated for AXG and G12A/G12B/SM1 BootROMs.

The main difference is that the TOC is in clear and located right
after the BL2.
For G12A/G12B/SM1 the BL2 is larger (64K instead of 48K).

### CODE REVIEW

File: fip.c [hunk 1]
```diff
@@ -330,9 +366,11 @@ struct fip_toc_info {
  *
  * @param fip: fill up with FIP informations
  * @param fd: FIP image to read header from
+ * @param rev: FIP Format revision
+ * @param bl2sz: BL2 Size
  * @return: 0 on success, negative number otherwise
  */
-static int fip_read_toc(struct fip_toc_info *toc, int fd)
+static int fip_read_toc(struct fip_toc_info *toc, int fd, enum fip_rev rev, size_t bl2sz)
```
Comment (repk) [inline fip.c]: Can you please do that in two lines in order to follow 80 characters line limit ?

File: fip.c [hunk 2]
```diff
@@ -130,6 +145,27 @@ static uuid_t const uuid_list[] = {
...
+	},
+	[FBI_BL32_DATA] = {
+		0x34, 0xa1, 0x48, 0xb8, 0xbc, 0x90, 0xe6, 0x11,
+                0x8f, 0xef, 0xa4, 0xba, 0xdb, 0x19, 0xde, 0x03
+	},
+	[FBI_BL33_DATA] = {
+		0x8e, 0x59, 0xd6 ,0x5d, 0x5e, 0x8b, 0xe6, 0x11,
+                0xbc, 0xb5, 0xf0, 0xde, 0xf1, 0x83, 0x72, 0x96,
+	},
+	[FBI_EMPTY] = {},
```
Comment (repk) [inline fip.c:168]: Are those only v3 related or are you aware of v2 format that can use them ?

**Review by repk — COMMENTED** [review]
First of all thanks a lot for this work this is really appreciated.

Sorry for the annoying coding style comments but I've seen you fixed that while I tried to get my hand on github review tool.

I have added a question about the new FIP UIDs, if in your experience you haven't seen any such FIP entries in v2 images I'll will merge this PR as is, on the other side if you think it can be used for v2 images as well I am wondering if it would be woth to put them in a separate commit.

**Comment by superna9999** [comment]
> First of all thanks a lot for this work this is really appreciated.

This is the first step of the work, I'm adding FIP creation at https://github.com/superna9999/gxlimg/tree/fip_v3

> Sorry for the annoying coding style comments but I've seen you fixed that while I tried to get my hand on github review tool.

No problem, I try to match the coding style, no problem to correct !

> I have added a question about the new TOC UIDs, if in your experience you haven't seen any such TOC entries in v2 images I'll will merge this PR as is, on the other side if you think it can be used for v2 images as well I am wondering if it would be woth to put them in a separate commit.

AFAIK I've only seen then on v3

File: fip.c (hunk 2 above)
Comment (superna9999) [reply fip.c:168]: AFAIK only v3 related, and no idea how this data is used, probably used when enabling secure boot somehow

**Comment by repk** [comment]
> > First of all thanks a lot for this work this is really appreciated.
> 
> This is the first step of the work, I'm adding FIP creation at https://github.com/superna9999/gxlimg/tree/fip_v3

Wow this is a pleasant surprise thanks a lot for the work. Maybe if FIP creation for v3 get merged at some point I would rename this as amlimg as stated here https://github.com/repk/gxlimg/issues/12.

**Comment by superna9999** [comment]
> > > First of all thanks a lot for this work this is really appreciated.
> > 
> > 
> > This is the first step of the work, I'm adding FIP creation at https://github.com/superna9999/gxlimg/tree/fip_v3
> 
> Wow this is a pleasant surprise thanks a lot for the work. Maybe if FIP creation for v3 get merged at some point I would rename this as amlimg as stated here #12.

Yep this work was triggered by https://groups.io/g/u-boot-amlogic/topic/87281853#1250 having an open source alternative would be much better if we pushed it in binman !

**Pull request merged**